prpl/attachments.c

```diff
@@ -192,3 +194,432 @@ void download_attachment(ChimeConnection *cxn, ChimeAttachment *att, AttachmentC
+
+	jb = json_builder_end_object(jb);
+
+	JsonNode *node = json_builder_get_root(jb);
+	JsonObject *obj = json_node_get_object(node);
+
+	chime_connection_send_message_async(
+		data->conn,
+		data->obj,
+		xfer->message,
+		NULL,
+		send_upload_confirmation_callback,
+		xfer,
+		obj
+	);
+
+	json_node_unref(node);
+	g_object_unref(jb);
+}
+
+static void put_file_callback(
+	SoupSession *session,
+	SoupMessage *msg,
+	gpointer user_data)
+{
+	purple_debug_misc("chime", "Put file request finished\n");
+	PurpleXfer *xfer = (PurpleXfer*)user_data;
+	AttachmentUpload *data = (AttachmentUpload*)xfer->data;
+
+	// This is freed by libsoup
+	data->soup_session = NULL;
+	data->soup_message = NULL;
+
+	if (purple_xfer_is_canceled(xfer)) return deep_free_upload_data(xfer);
+
+	if (!SOUP_STATUS_IS_SUCCESSFUL(msg->status_code)) {
+		gchar *error_msg = g_strdup_printf(_("Failed to upload file: (%d) %s"), msg->status_code, msg->reason_phrase);
+		purple_debug_misc("chime", "%s\n", error_msg);
+		purple_xfer_conversation_write(xfer, error_msg, TRUE);
+		g_free(error_msg);
+		deep_free_upload_data(xfer);
+		return;
+	}
+
+	const char *etag;
+	etag = soup_message_headers_get_one(msg->response_headers, "ETag");
+	purple_debug_misc("chime", "Extracted ETag: %s\n", etag);
+
+	if (!etag) {
+		purple_debug_error("chime", "Could not extract ETag value from HTTP headers\n");
+		deep_free_upload_data(xfer);
+		return;
+	}
+
+	// We need to send a message confirming the upload
+	send_upload_confirmation(xfer, etag);
+}
+
+static void update_progress(
+	SoupMessage *msg,
+	SoupBuffer *chunk,
+	gpointer user_data)
+{
+	PurpleXfer *xfer = (PurpleXfer*)user_data;
+	xfer->bytes_sent = xfer->bytes_sent + chunk->length;
+	xfer->bytes_remaining = xfer->bytes_remaining - chunk->length;
+	purple_debug_misc("chime", "Updating progress by %lu bytes. Sent=%lu, Remaining=%lu\n",
+			chunk->length, xfer->bytes_sent, xfer->bytes_remaining);
+	purple_xfer_update_progress(xfer);
+}
+
+static void put_file(ChimeConnection *cxn, PurpleXfer *xfer)
+{
+	purple_debug_misc("chime", "Submitting put file request\n");
+
+	AttachmentUpload *data = (AttachmentUpload*)xfer->data;
+	gchar *content_length = g_strdup_printf("%lu", data->content_length);
+
+	SoupMessage *msg;
+	data->soup_message = msg = soup_message_new("PUT", data->upload_url);
+
+	soup_message_set_request(msg, data->content_type,
+			SOUP_MEMORY_TEMPORARY, data->content, data->content_length);
+	soup_message_headers_append(msg->request_headers, "Cache-Control", "no-cache");
+	soup_message_headers_append(msg->request_headers, "Pragma", "no-cache");
+	soup_message_headers_append(msg->request_headers, "Accept", "*/*");
+	soup_message_headers_append(msg->request_headers, "Content-length", content_length);
+
+	g_signal_connect(msg, "wrote-body-data", (GCallback)update_progress, xfer);
+
+	data->soup_session = soup_session_new_with_options(
+			SOUP_SESSION_ADD_FEATURE_BY_TYPE,
+			SOUP_TYPE_CONTENT_SNIFFER, // TODO: Check this type
```

Comment:
Not sure which TYPE should be used here. Tried a few and it didn't make much difference.